### PR TITLE
Rewrite call state machine

### PR DESCRIPTION
**Pull request by dbkr**
* Remove the two separate enumerations of call state: now everything
   uses the js-sdk version of call state. Stop adding a separate
   'call_state' field onto the call object(!)
 * Better reflection of the actual state of the call in the call bar,
   so when it's connecting, it says connecting, and only says 'active call'
   when the call is actually active.
 * More typey goodness

Requires https://github.com/matrix-org/matrix-js-sdk/pull/1503

### CODE REVIEW

File: src/CallHandler.tsx [hunk 1]
```diff
@@ -77,13 +77,18 @@ import ErrorDialog from "./components/views/dialogs/ErrorDialog";
 import WidgetStore from "./stores/WidgetStore";
 import { WidgetMessagingStore } from "./stores/widgets/WidgetMessagingStore";
 import { ElementWidgetActions } from "./stores/widgets/ElementWidgetActions";
+import { MatrixCall, CallErrorCode, CallState, CallType } from "matrix-js-sdk/lib/webrtc/call";
 
-// until we ts-ify the js-sdk voip code
-type Call = any;
+enum AudioId {
```
Comment (turt2live) [inline src/CallHandler.tsx:82]: `AudioID` to match conventions (which are informal and awful, so happy to be argued on this)

File: src/CallHandler.tsx [hunk 2]
```diff
@@ -185,69 +190,57 @@ export default class CallHandler {
...
-                this.setCallState(call, call.roomId, "ringback");
-                this.play("ringbackAudio");
-            } else if (newState === "ended" && oldState === "connected") {
-                this.removeCallForRoom(call.roomId);
-                this.pause("ringbackAudio");
-                this.play("callendAudio");
-            } else if (newState === "ended" && oldState === "invite_sent" &&
-                    (call.hangupParty === "remote" ||
-                    (call.hangupParty === "local" && call.hangupReason === "invite_timeout")
+        call.on("state", (newState: CallState, oldState: CallState) => {
```
Comment (turt2live) [inline src/CallHandler.tsx:193]: ideally "state" would also be enumified

File: src/CallHandler.tsx [hunk 3]
```diff
@@ -185,69 +190,57 @@ export default class CallHandler {
...
+                    this.play(AudioId.Ring);
+                    break;
+                case CallState.InviteSent:
+                    this.play(AudioId.Ringback);
+                    break;
+                case CallState.Ended:
+                    this.removeCallForRoom(call.roomId);
+                    if (oldState === CallState.InviteSent && (
+                        call.hangupParty === "remote" ||
+                        (call.hangupParty === "local" && call.hangupReason === "invite_timeout")
```
Comment (turt2live) [inline src/CallHandler.tsx:216]: if these things are protocol values we can rely on, enumification :)

File: src/CallHandler.tsx [hunk 4]
```diff
@@ -279,36 +272,36 @@ export default class CallHandler {
...
-                    payload.local_element,
-                );
-            } else {
-                console.error("Unknown conf call type: %s", payload.type);
+
+    private placeCall(roomId: string, type: CallType, localElement: HTMLVideoElement, remoteElement: HTMLVideoElement) {
+        const call = Matrix.createNewMatrixCall(MatrixClientPeg.get(), roomId);
+        this.calls.set(roomId, call);
+        this.setCallListeners(call);
+        if (type === 'voice') {
```
Comment (turt2live) [inline src/CallHandler.tsx:280]: enum usage

File: src/CallHandler.tsx [hunk 5]
```diff
@@ -279,36 +272,36 @@ export default class CallHandler {
...
+            call.placeVoiceCall();
+        } else if (type === 'video') {
```
Comment (turt2live) [inline src/CallHandler.tsx:282]: enum usage

File: src/CallHandler.tsx [hunk 6]
```diff
@@ -279,36 +272,36 @@ export default class CallHandler {
...
+            call.placeVideoCall(
+                remoteElement,
+                localElement,
+            );
+        } else if (type === 'screensharing') {
```
Comment (turt2live) [inline src/CallHandler.tsx:287]: enum usage

File: src/components/structures/RoomStatusBar.js [hunk 7]
```diff
@@ -121,6 +121,10 @@ export default class RoomStatusBar extends React.Component {
         });
     };
 
+    _showCallBar() {
+        return this.props.callState !== 'ended' && this.props.callState !== 'ringing';
```
Comment (turt2live) [inline src/components/structures/RoomStatusBar.js:125]: should be able to use enums here too

File: src/components/structures/RoomStatusBar.js [hunk 8]
```diff
@@ -269,6 +273,25 @@ export default class RoomStatusBar extends React.Component {
         </div>;
     }
 
+    _getCallStatusText() {
+        switch (this.props.callState) {
```
Comment (turt2live) [inline src/components/structures/RoomStatusBar.js:277]: enum the cases please :)

File: src/components/structures/RoomView.tsx [hunk 9]
```diff
@@ -1742,10 +1738,12 @@ export default class RoomView extends React.Component<IProps, IState> {
...
-        let inCall = false;
-        if (call && (this.state.callState !== 'ended' && this.state.callState !== 'ringing')) {
-            inCall = true;
+        let activeCall = null;
+        {
+            const call = this.getCallForRoom();
+            if (call && (this.state.callState !== 'ended' && this.state.callState !== 'ringing')) {
+                activeCall = call;
+            }
         }
```
Comment (turt2live) [inline src/components/structures/RoomView.tsx:1747]: no need for the block?

File: src/components/structures/RoomView.tsx [hunk 10]
```diff
@@ -1890,10 +1889,10 @@ export default class RoomView extends React.Component<IProps, IState> {
             };
         }
 
-        if (inCall) {
+        if (activeCall) {
             let zoomButton; let videoMuteButton;
 
-            if (call.type === "video") {
+            if (activeCall.type === "video") {
```
Comment (turt2live) [inline src/components/structures/RoomView.tsx:1895]: enuuummmm

**Review by turt2live — CHANGES_REQUESTED** [review]
generally lgtm, though most of my review can be summarized as: enum

File: src/components/structures/RoomView.tsx (hunk 9 above)
Comment (dbkr) [reply src/components/structures/RoomView.tsx:1747]: I just didn't want the 'call' variable hanging around for the rest of the function

Comment (turt2live) [reply src/components/structures/RoomView.tsx:1747]: ah, would be good to comment on the block then

**Review by turt2live — APPROVED** [review]
lgtm - would be good to get a comment on the block just because it's a relatively unexpected thing in our codebase

**Pull request merged**